improver/regrid/grid.py

```diff
@@ -0,0 +1,395 @@
+
+    Args:
+        cube_in_mask (iris.cube.Cube):
+            land_sea mask information cube for input source grid(land=>1)
+            should in GeogCS's lats/lons coordinate system
+        classify_latlons(numpy.ndarray):
+            latitude and longitude source grid points to classify (N x 2)
+
+    Returns:
+        numpy.ndarray: classifications (N) for 1D-ordered source grid points
+    """
+    in_land_mask = cube_in_mask.data
+    lats_name, lons_name = latlon_names(cube_in_mask)
+    in_land_mask_lats = cube_in_mask.coord(lats_name).points
+    in_land_mask_lons = cube_in_mask.coord(lons_name).points
+
+    mask_rg_interp = RegularGridInterpolator(
+        (in_land_mask_lats, in_land_mask_lons),
+        in_land_mask,
+        method="nearest",
+        bounds_error=False,
+        fill_value=0.0,
+    )
+    is_land = np.bool_(mask_rg_interp(classify_latlons))
+    return is_land
+
+
+def slice_cube_by_domain(cube_in, output_domain):
+    """
+    extract cube domain to be consistent as cube_reference's domain
+
+    Args:
+        cube_in (iris.cube.Cube):
+            input data cube to be sliced
+        output_domain(tuple):
+            lat_max,lon_max,lat_min,lon_min
+
+    Returns:
+        cube_in (iris.cube.Cube):
+            data cube after slicing
+    """
+    lat_max, lon_max, lat_min, lon_min = output_domain
+    lat_d, lon_d = get_grid_spacing(cube_in)
+
+    domain = iris.Constraint(
+        latitude=lambda val: lat_min - 2.0 * lat_d < val < lat_max + 2.0 * lat_d
+    ) & iris.Constraint(
+        longitude=lambda val: lon_min - 2.0 * lon_d < val < lon_max + 2.0 * lon_d
+    )
+
+    cube_in = cube_in.extract(domain)
+
+    return cube_in
+
+
+def slice_mask_cube_by_domain(cube_in, cube_in_mask, output_domain):
+    """
+    extract cube domain to be consistent as cube_reference's domain
+
+    Args:
+        cube_in (iris.cube.Cube):
+            input data cube to be sliced
+        cube_in_mask (iris.cube.Cube):
+            input maskcube to be sliced
+        output_domain (Tuple[float, float, float, float]):
+            lat_max, lon_max, lat_min, lon_min
+    Returns:
+        Tuple[iris.cube.cube, iris.cube.Cube]:
+            data cube after slicing, mask cube after slicing
+    """
+    lat_max, lon_max, lat_min, lon_min = output_domain
+    lat_d_1, lon_d_1 = get_grid_spacing(cube_in)
+    lat_d_2, lon_d_2 = get_grid_spacing(cube_in_mask)
+    lat_d = lat_d_1 if lat_d_1 > lat_d_2 else lat_d_2
+    lon_d = lon_d_1 if lon_d_1 > lon_d_2 else lon_d_2
+
+    domain = iris.Constraint(
+        latitude=lambda val: lat_min - 2.0 * lat_d < val < lat_max + 2.0 * lat_d
+    ) & iris.Constraint(
+        longitude=lambda val: lon_min - 2.0 * lon_d < val < lon_max + 2.0 * lon_d
+    )
+
+    cube_in = cube_in.extract(domain)
+    cube_in_mask = cube_in_mask.extract(domain)
+
+    return cube_in, cube_in_mask
+
+
+def create_regrid_cube(cube_array, cube_in, cube_out):
```

Comment:
We have several utilities for creating new cubes within IMPROVER, but they all assume the new cube we want to create has the same x-y dimensions as one of the input cubes. I need to go back and check, and maybe chat to someone else from the Met Office, about whether any of these could be reused here to help ensure consistency.

Reply:
This function has some special requirement here, as we need cube_in for some coord  and other information.
In the new code, we convert target's curvilinear system into lat/lon first, and then go back to original system. 
Please check if the function is consistent with your requirement. BTW, BOM is going to use Albers equal-area projection, I reckon that the new code can handle it, and  I will do some testing soon.

Reply:
I've chatted this through with some people at the Met Office and we agree that this function has the correct behaviour. But I've put a couple of comments below about how it can be done neater/clearer.